utils/uftrace-script.h

```diff
@@ -0,0 +1,54 @@
+#ifndef UFTRACE_UFTRACE_SCRIPT_H
+#define UFTRACE_UFTRACE_SCRIPT_H
+
+#ifndef __cplusplus
+#include <stdint.h>
+#include <stdbool.h>
+#include "utils/list.h"
+#else
+#include <cstdint>
+#include <cstdbool>
+#endif
+
+/* informantion passed during initialization */
+#define SCRIPT_INFO_MAJOR_VERSION 1
+#define SCRIPT_INFO_MINOR_VERSION 0
+struct script_info {
+	struct si_version {
+		uint8_t major;
+		uint8_t minor;
+	} ver;
+	char			*name;
+	char			*version;
+	bool			record;
+	const char 		*cmds;
+};
+
+/* context information passed to script */
+#define SCRIPT_CONTEXT_MAJOR_VERSION 1
+#define SCRIPT_CONTEXT_MINOR_VERSION 0
+struct script_context {
+	struct sc_version {
+		uint8_t major;
+		uint8_t minor;
+	} ver;
+	int			tid;
+	int			depth;
+	uint64_t		timestamp;
+	uint64_t		duration;	/* exit only */
+	unsigned long		address;
+	char			*name;
+#ifndef __cplusplus
+	/* for arguments and return value */
+	int			arglen;
+	void			*argbuf;
+	struct list_head	*argspec;
+#else
+	/*
+	 * struct list_head internally uses 'new' as a variable name so
+	 * it's not possible to use in C++.
+	 */
+#endif
```

Comment:
We may have to convert the `struct list_head` to this type. Could you explain why you used `char data[]` here?

Otherwise, how about modifying `utils/list.h` that can be compilable in C++, then distribute it together because uftrace also uses GPLv2 same as `list.h` anyway.

Reply:
Thanks for the suggestion. I see that if we introduce another way of handling arguments then we better use this in `script-python.c` and `script-luajit.c` as well without using the current `struct list_head`.

It may take more time to refine them all so I didn't update it for now.